src/lib/x509/x509_ext.cpp

```diff
@@ -868,6 +877,607 @@ const std::string& TNAuthList::Entry::telephone_number() const {
    return std::get<ASN1_String>(m_data).value();
 }
 
+std::vector<uint8_t> IPAddressBlocks::encode_inner() const {
+   std::vector<uint8_t> output;
+   DER_Encoder(output).start_sequence().encode_list(m_ip_addr_blocks).end_cons();
+   return output;
+}
+
+void IPAddressBlocks::decode_inner(const std::vector<uint8_t>& in) {
+   BER_Decoder(in).decode_list(m_ip_addr_blocks).verify_end();
+}
+
+void IPAddressBlocks::IPAddressFamily::encode_into(Botan::DER_Encoder& into) const {
+   into.start_sequence();
+
+   std::vector<uint8_t> afam = {static_cast<uint8_t>(m_afi >> 8), static_cast<uint8_t>(m_afi & 0xff)};
+
+   if(m_safi.has_value()) {
+      afam.push_back(m_safi.value());
+   }
+
+   into.add_object(ASN1_Type::OctetString, ASN1_Class::Universal, afam);
+
+   if(std::holds_alternative<IPAddressChoice<Version::IPv4>>(m_ip_addr_choice)) {
+      into.encode(std::get<IPAddressChoice<Version::IPv4>>(m_ip_addr_choice));
+   } else {
+      into.encode(std::get<IPAddressChoice<Version::IPv6>>(m_ip_addr_choice));
+   }
+   into.end_cons();
+}
+
+void IPAddressBlocks::IPAddressFamily::decode_from(Botan::BER_Decoder& from) {
+   const BER_Object obj = from.get_next_object();
+   const ASN1_Type type_tag = obj.type_tag();
+   if(type_tag != ASN1_Type::Sequence) {
+      throw Decoding_Error(fmt("Unexpected type for IPAddressFamily {}", static_cast<uint32_t>(type_tag)));
+   }
```

Comment:
You can use `peek_next_object` here and avoid having to create a second `BER_Decoder`